File: extractors/typescript/src/stats.ts
/**
 * The stderr summary, in the Java and C# extractors' format. The rate
 * measures the corpus's DEPENDENCY SURFACE (PLAN.md §5.3): a package that is
 * not under the roots cannot be resolved by any extractor, so a low rate is a
 * fact about the corpus, and the stub discipline is the property worth
 * asserting. The `any`-receiver count is this profile's stated ceiling.
 */
export class ResolutionStats {
  typeReferences = 0;
  unresolved = 0;
  imports = 0;
  importsUnresolved = 0;
  importsWorkspaceResolved = 0;
  selfEdgesDropped = 0;
  anyReceiversDropped = 0;
  computedAccessesDropped = 0;
  computedImportsDropped = 0;
  heritageExpressionsDropped = 0;
  namelessClassesDropped = 0;
  /** Same-keyed declarations in one file: the later one re-keyed by its kind or position. */
  readonly duplicateKeys: string[] = [];
  /** Stubs by origin, for the summary line. */
  stubs = { lib: 0, packages: 0, unresolved: 0, modules: 0 };

  get resolved(): number {
    return this.typeReferences - this.unresolved;
  }

  noteTypeReference(resolved: boolean): void {
    this.typeReferences += 1;
    if (!resolved) this.unresolved += 1;
  }

  noteImport(outcome: "resolved" | "unresolved" | "workspace"): void {
    this.imports += 1;
    if (outcome === "unresolved") this.importsUnresolved += 1;
    if (outcome === "workspace") this.importsWorkspaceResolved += 1;
  }

  summary(entities: number, stubs: number, edges: number): string {
    const rate = this.typeReferences === 0 ? 100 : (100 * this.resolved) / this.typeReferences;
    return [
      "RESOLUTION SUMMARY",
      `  type references : ${this.typeReferences}`,
      `  resolved        : ${this.resolved}`,
      `  unresolved      : ${this.unresolved}`,
      `  resolution rate : ${rate.toFixed(1)}%`,
      `  any-typed receivers (dropped) : ${this.anyReceiversDropped}`,
      `  imports         : ${this.imports} (unresolved: ${this.importsUnresolved}, workspace-resolved: ${this.importsWorkspaceResolved})`,
      `  entities        : ${entities} (stubs: ${stubs} — lib ${this.stubs.lib}, packages ${this.stubs.packages}, <unresolved> ${this.stubs.unresolved}, modules ${this.stubs.modules})`,
      `  edges           : ${edges} (self-edges dropped: ${this.selfEdgesDropped}, computed accesses dropped: ${this.computedAccessesDropped}, computed imports dropped: ${this.computedImportsDropped}, heritage expressions dropped: ${this.heritageExpressionsDropped}, nameless classes dropped: ${this.namelessClassesDropped})`,
      `  duplicates      : ${this.duplicateKeys.length} same-keyed declarations re-keyed (first in file order keeps the plain key)`,
      "",
    ].join("\n");
  }
}

File: extractors/elixir/lib/codegraph_elixir/stats.ex
defmodule CodegraphElixir.Stats do
  @moduledoc """
  The stderr summary, in the other extractors' format. Every count here is a
  fact the model cannot carry: what was dropped and why, so the resolution
  ceiling of the baseline is stated rather than hidden (PLAN.md §16.5).

  `references` are the sites the walker read; `resolved` those that became
  an edge (to a corpus entity or to a stub — resolvability is not
  membership); `kernel` those that bound to the language itself (never an
  edge); everything else is dropped under its reason.
  """

  defstruct files_ex: 0,
            files_exs: 0,
            unparsed: [],
            modules: 0,
            imports: %{alias: 0, import: 0, require: 0, use: 0},
            imports_unresolved: 0,
            references: 0,
            resolved: 0,
            external: 0,
            kernel: 0,
            emitted: %{},
            dropped: %{},
            self_edges_dropped: 0,
            dynamic_modules_dropped: 0,
            duplicate_keys: [],
            duplicate_modules: [],
            unclosable: [],
            stubs: %{otp: 0, deps: 0}

  def drop(%__MODULE__{} = s, reason),
    do: %__MODULE__{s | dropped: Map.update(s.dropped, reason, 1, &(&1 + 1))}

  def drop(%__MODULE__{} = s, reason, n),
    do: %__MODULE__{s | dropped: Map.update(s.dropped, reason, n, &(&1 + n))}

  def summary(%__MODULE__{} = s, model) do
    stubs = s.stubs.otp + s.stubs.deps
    attempted = s.references
    rate = if attempted == 0, do: 100.0, else: 100.0 * (s.resolved + s.kernel) / attempted
    imports = s.imports.alias + s.imports.import + s.imports.require + s.imports.use
    dropped_total = s.dropped |> Map.values() |> Enum.sum()

    dropped =
      s.dropped
      |> Enum.sort_by(fn {reason, _} -> Atom.to_string(reason) end)
      |> Enum.map_join(", ", fn {reason, n} -> "#{reason} #{n}" end)

    emitted =
      s.emitted
      |> Enum.sort_by(fn {kind, _} -> kind end)
      |> Enum.map_join(", ", fn {kind, n} -> "#{kind} #{n}" end)

    """
    RESOLUTION SUMMARY
      references      : #{attempted}
      resolved        : #{s.resolved} (external: #{s.external}) + kernel: #{s.kernel} (the language, never an edge)
      dropped         : #{dropped_total}#{if dropped == "", do: "", else: " (" <> dropped <> ")"}
      resolution rate : #{:erlang.float_to_binary(rate, decimals: 1)}%
      imports         : #{imports} (alias #{s.imports.alias}, import #{s.imports.import}, require #{s.imports.require}, use #{s.imports.use}; external: #{s.imports_unresolved})
      files           : #{s.files_ex + s.files_exs} (.ex #{s.files_ex}, .exs #{s.files_exs}; unparsed: #{length(s.unparsed)})
      entities        : #{length(model.entities)} (stubs: #{stubs} — <otp> #{s.stubs.otp}, <deps> #{s.stubs.deps})
      edges           : #{length(model.edges)}#{if emitted == "", do: "", else: " (" <> emitted <> ")"}; self-edges dropped: #{s.self_edges_dropped}, dynamic module names dropped: #{s.dynamic_modules_dropped}
      duplicates      : #{length(s.duplicate_keys)} same-keyed declarations re-keyed, #{length(s.duplicate_modules)} module names declared in more than one file (the first in file order is the import target)
      unclosable      : #{length(s.unclosable)} edges dropped (an endpoint no entity declares)
    """
  end
end
